utils/contiguous_shared_buffer.hh

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <concepts>
+#include <memory>
+
+// A contiguous buffer of char objects which can be trimmed and
+// supports zero-copy sharing of its underlying memory.
+template<typename T>
+concept ContiguousSharedBuffer = std::movable<T>
+    && std::default_initializable<T>
+    && requires(T& obj, size_t pos, size_t len) {
+
+        // Creates a new buffer that shares the memory of the original buffer.
+        // The lifetime of the new buffer is independent of the original buffer.
+        { obj.share() } -> std::same_as<T>;
+
+        // Like share() but the new buffer represents a sub-range of the original buffer.
+        { obj.share(pos, len) } -> std::same_as<T>;
+
```

Comment:
Should we remove share() from the buffer provided by the cache, since it's unsafe?

Reply:
This `share` is safe. The shareable thing is `page_view` itself, which acts as a specialized `shared_ptr` to the underlying `cached_page` object (allocated in the standard allocator), which holds the pointer (updated on LSA migrations) to the actual LSA data.

It's fine to store and share the `page_view`. You only can't keep the `const char*` it returns (via `page_view::begin()`) across LSA reclaim points — after every reclaim point you have to get the current value via `begin()`.

Reply:
share() for page_view is safe, because it returns another page_view. What's unsafe is share() on temporary_buffer which is a view of some snapshot of page_view - that gets invalidated by LSA moves. page_view is not invalidated by moves, it follows to buffer.

That's the approach taken here, to keep share() safe for both, by passing page_view directly to parsers.